### ISSUE TITLE
Read-Write Splitting

### ISSUE THREAD

**Issue by stevefan1999-personal** [ExpediaGroup/graphql-kotlin#881]
tl;dr: The separation made by GraphQL, i.e. Query/Mutation/Subscription makes it relatively easy to do read write separation, where Query corresponds to parallelizable read-replications, and Mutation correspond to (most likely serializable) write. With current state of modern database HA design, together we can leverage this to get a huge performance gain.

However, it seems like graphql-kotlin does not allow us to do this out of the box, and specifically required us to mark each and every function manually to use the different data source respectively. This would be verbosive, being a huge boilerplate and has a great code smell. I'm currently using graphql-kotlin in conjunction with Spring.

My idea is very simple: We should be able to set different data source on GraphQLContextFactory, e.g. we can detect whether our next resolver request is a Query/Mutation and we can tell Spring to dynamically switch to read replication/master pool respectively.

**Comment by dariuszkuc** [comment]
Hello! Thanks for raising this but I don't think we will be implementing this in any near future as you described a very specific use case that is not really generalizable. GraphQL is not synonymous with database access nor it has any knowledge of any transactional data - it is up to the developers to implement that in their resolvers. 

On a side note -> I would generally not recommend to do this heavy lifting in the context factory as it is executed before HTTP body is parsed so in order to determine what sort of operations user requested you would have to explicitly parse it. Additionally, if you are using GraphQL to expose your database schema I'd highly recommend to abstract it away and instead expose "client" schema. I'd also argue that if you are ever in a situation where you need to execute multiple mutations in single transactions then probably your schema should be rewritten to expose abstracted single operation.
